pkg/storage/chunk/client/gcp/gcs_thanos_object_client.go

```diff
@@ -0,0 +1,203 @@
+	"google.golang.org/api/googleapi"
+	amnet "k8s.io/apimachinery/pkg/util/net"
+
+	"github.com/grafana/loki/v3/pkg/storage/bucket"
+	"github.com/grafana/loki/v3/pkg/storage/chunk/client"
+	"github.com/grafana/loki/v3/pkg/storage/chunk/client/hedging"
+)
+
+type GCSThanosObjectClient struct {
+	client       objstore.Bucket
+	hedgedClient objstore.Bucket
+	logger       log.Logger
+}
+
+func NewGCSThanosObjectClient(ctx context.Context, cfg bucket.Config, component string, logger log.Logger, hedgingCfg hedging.Config) (*GCSThanosObjectClient, error) {
+	client, err := newGCSThanosObjectClient(ctx, cfg, component, logger, false, hedgingCfg)
+	if err != nil {
+		return nil, err
+	}
+
+	hedgedClient, err := newGCSThanosObjectClient(ctx, cfg, component, logger, true, hedgingCfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return &GCSThanosObjectClient{
+		client:       client,
+		hedgedClient: hedgedClient,
+		logger:       logger,
+	}, nil
+}
+
+func newGCSThanosObjectClient(ctx context.Context, cfg bucket.Config, component string, logger log.Logger, hedging bool, hedgingCfg hedging.Config) (objstore.Bucket, error) {
+	if hedging {
+		hedgedTrasport, err := hedgingCfg.RoundTripperWithRegisterer(nil, prometheus.WrapRegistererWithPrefix("loki_", prometheus.DefaultRegisterer))
+		if err != nil {
+			return nil, err
+		}
+
+		cfg.GCS.Transport = hedgedTrasport
+	}
+	return bucket.NewClient(bucket.GCS, ctx, cfg, component, logger)
+}
+
+func (s *GCSThanosObjectClient) Stop() {
+}
+
+// ObjectExists checks if a given objectKey exists in the GCS bucket
+func (s *GCSThanosObjectClient) ObjectExists(ctx context.Context, objectKey string) (bool, error) {
+	return s.client.Exists(ctx, objectKey)
+}
+
+// GetAttributes returns the attributes of the specified object key from the configured GCS bucket.
+func (s *GCSThanosObjectClient) GetAttributes(ctx context.Context, objectKey string) (client.ObjectAttributes, error) {
+	attr := client.ObjectAttributes{}
+	thanosAttr, err := s.hedgedClient.Attributes(ctx, objectKey)
+	if err != nil {
+		return attr, err
+	}
+	attr.Size = thanosAttr.Size
+
+	return attr, nil
+}
+
+// PutObject puts the specified bytes into the configured GCS bucket at the provided key
+func (s *GCSThanosObjectClient) PutObject(ctx context.Context, objectKey string, object io.Reader) error {
+	return s.client.Upload(ctx, objectKey, object)
+}
+
+// GetObject returns a reader and the size for the specified object key from the configured GCS bucket.
+// size is set to -1 if it cannot be succefully determined, it is up to the caller to check this value before using it.
+func (s *GCSThanosObjectClient) GetObject(ctx context.Context, objectKey string) (io.ReadCloser, int64, error) {
+	reader, err := s.hedgedClient.Get(ctx, objectKey)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	size, err := objstore.TryToGetSize(reader)
+	if err != nil {
+		size = -1
+		level.Warn(s.logger).Log("msg", "failed to get size of object", "err", err)
+	}
```

Comment:
likely not for gcs, but it might also fail if the provider specific code fails to get the size for some reason. Seems very unlikely but I added this to not fail the Get request entirely if we cannot figure out the size. Size is only being used to reduce allocations, does not affect any functionality